### ISSUE TITLE
onComplete being called on every render

### ISSUE THREAD

**Issue by vititu** [apollographql/react-apollo#2522]
The changes made related to issue #2177 causes `onCompleted` to be called on every render. Before it was called once.

That might break the app causing a `Maximum update depth exceeded error` if trying to setSate with the data provided.

**Comment by JoviDeCroock** [comment]
Why would you use `setState` with provided data, just wondering.

**Comment by geekplux** [comment]
same problem.

**Comment by riccoski** [comment]
`I think the best way is to use the skip prop.
onCompleted() {

setState({
    skip: true
});
<Query skip={this.state.skip>
`

**Comment by asgeirfri** [comment]
Does anyone know the propper way to only setState after server call? Thanks.

**Comment by charlex** [comment]
This is happening to me as well. 

This seems to be functioning differently from what I understand the intended behavior of `onCompleted` to be.

Any hints as to why this is happening?

**Comment by JaosnHsieh** [comment]
you can add an if else condition in completed function in order to prevent what will change state again and again.

```javascript
       <Query
            query={XXXXX}
            onCompleted={(data) => {
              if (!initValue) {
                   ...do something
              }
            }}
          >
```

**Comment by hwillson** [comment]
This has been fixed in current day versions of `react-apollo`. Thanks!

**Comment by charlex** [comment]
That's great! @hwillson Thank you for updating this ticket!

**Comment by kcvin94** [comment]
still happening actually...

**Comment by dylanwulf** [comment]
@kcvin94 Should be fixed in the newly released v3.1.0. See #3353 for more details.
